### PR TITLE
platform(general): Handle SAST suppressions (suppressions V2)

### PR DESCRIPTION
**Pull request by mikeurbanski1**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

[//]: # "
    # PR Title
    Be aware that we use the title to create changelog automatically and therefore only allow specific prefixes
    - break:    to indicate a breaking change, this supersedes any of the types
    - feat:     to indicate new features or checks
    - fix:      to indicate a bugfix or handling of edge cases of existing checks
    - docs:     to indicate an update to our documentation
    - chore:    to indicate adjustments to workflow files or dependency updates
    - platform: to indicate a change needed for the platform
    Additionally a scope is needs to be added to the prefix, which indicates the targeted framework, in doubt choose 'general'.
    #    
    Allowed prefixs:
    ansible|argo|arm|azure|bicep|bitbucket|circleci|cloudformation|dockerfile|github|gha|gitlab|helm|kubernetes|kustomize|openapi|sast|sca|secrets|serverless|terraform|general|graph|terraform_plan|terraform_json
    #
    ex.
    feat(terraform): add CKV_AWS_123 to ensure that VPC Endpoint Service is configured for Manual Acceptance
"

## Description

- Adds logic to get and apply suppressions V2 (currently SAST) for file, repo, and policy suppressions. Findings cannot be suppressed due to no way to map a finding ID in a suppression to a checkov finding.
- Adds logic to save policy suppressions for use in filtering out checks in the runner filter and handling cloned policies

Also adds logic that is somewhat related (still related to customizations) to pass sast check metadata (severities) to sast core, so that the right severity gets used. See also https://code.pan.run/bridgecrew/sast-core/-/merge_requests/655

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my feature, policy, or fix is effective and works
- [X] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/common/bridgecrew/integration_features/features/suppressions_integration.py [hunk 1]
```diff
@@ -106,17 +141,22 @@ def _apply_suppressions_to_report(self, scan_report: Report) -> None:
         scan_report.failed_checks = still_failed_checks
         scan_report.passed_checks = still_passed_checks
 
-    def _check_suppressions(self, record: Record, suppressions: list[dict[str, Any]]) -> dict[str, Any] | None:
+    def _check_suppressions(self, record: Record, suppressions: Optional[list[dict[str, Any]]], suppressions_v2: Optional[list[dict[str, Any]]]) -> dict[str, Any] | None:
         """
-        Checks the specified suppressions against the specified record, returning the first applicable suppression,
-        or None of no suppression is applicable.
+        Checks the specified suppressions against the specified record, returning a tuple of the applied suppression and whether
+        the suppression was a v1 suppression or not. If no suppression is found, then returns a tuple of None, None
```
Comment (tronxd) [inline checkov/common/bridgecrew/integration_features/features/suppressions_integration.py:147]: by the function type hints it looks like that the function returns a list of dictionaries, not tuples. what should it be?

File: checkov/common/bridgecrew/integration_features/features/suppressions_integration.py [hunk 2]
```diff
@@ -186,6 +226,30 @@ def _check_suppression(self, record: Record, suppression: dict[str, Any]) -> boo
...
+        if type == 'policy':  # TODO policy suppression not supported via UI yet but we just need to set the correct type value here
+            # We just checked the policy ID above
+            return True
+        elif type == 'finding':
+            pass  # TODO how to map them?
+        elif type == 'file':
+            record_file_path = record.repo_file_path if record.repo_file_path.startswith('/') else f'/{record.repo_file_path}'
+            for file_suppression in suppression['files']:
+                suppression_file_path = file_suppression['filePath']
+                suppression_file_path = suppression_file_path if suppression_file_path.startswith('/') else f'/{suppression_file_path}'
```
Comment (tronxd) [inline checkov/common/bridgecrew/integration_features/features/suppressions_integration.py:244]: please put the `/` prefix validation and prepend in a function

File: checkov/common/bridgecrew/integration_features/features/suppressions_integration.py [hunk 3]
```diff
@@ -186,6 +226,30 @@ def _check_suppression(self, record: Record, suppression: dict[str, Any]) -> boo
...
+                if self.bc_integration.repo_matches(file_suppression['repositoryName']) \
+                        and (suppression_file_path == record_file_path
+                             or suppression_file_path == convert_to_unix_path(record_file_path)):
```
Comment (tronxd) [inline checkov/common/bridgecrew/integration_features/features/suppressions_integration.py:247]: please refactor the condition into a function to understand what we check for

**Review by tronxd — APPROVED** [review]

File: checkov/common/bridgecrew/integration_features/features/suppressions_integration.py (hunk 1 above)
Comment (mikeurbanski1) [reply checkov/common/bridgecrew/integration_features/features/suppressions_integration.py:147]: Good catch, this was outdated from an earlier version of the method I decided to change

**Review by lirshindalman — APPROVED** [review]

File: checkov/common/bridgecrew/integration_features/features/suppressions_integration.py [hunk 4]
```diff
@@ -106,17 +141,19 @@ def _apply_suppressions_to_report(self, scan_report: Report) -> None:
...
-        :param record:
-        :param suppressions:
+        Checks the specified suppressions against the specified record, returning the applied suppression, if any, else None
         :return:
         """
-        for suppression in suppressions:
-            if self._check_suppression(record, suppression):
-                return suppression
+        if suppressions:
+            for suppression in suppressions or []:
```
Comment (lirshindalman) [inline checkov/common/bridgecrew/integration_features/features/suppressions_integration.py:150]: why do you use the 'or []' if you have 'if' before?

File: checkov/common/bridgecrew/integration_features/features/suppressions_integration.py [hunk 5]
```diff
@@ -186,6 +223,38 @@ def _check_suppression(self, record: Record, suppression: dict[str, Any]) -> boo
...
+        if type == 'policy':
+            # We just checked the policy ID above
+            return True
+        elif type == 'finding':
+            pass  # TODO how to map them?
+        elif type == 'file':
+            record_file_path = SuppressionsIntegration.normalize_file_path(record.repo_file_path)
+            for file_suppression in suppression['files']:
+                suppression_file_path = SuppressionsIntegration.normalize_file_path(file_suppression['filePath'])
+                if self._check_suppression_v2_file(record_file_path, suppression_file_path, file_suppression['repositoryName']):
```
Comment (achiar99) [inline checkov/common/bridgecrew/integration_features/features/suppressions_integration.py:252]: For old record we dont have `repositoryName` so lets use `file_suppression.get('repositoryName')`

**Review by achiar99 — APPROVED** [review]

**Pull request merged**